### PR TITLE
Feature/patient reserve appointment

### PR DESCRIPTION
**Pull request by mohamedebrahim2020**

### CODE REVIEW

File: app/Filters/AppointmentFilters.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+use App\Enums\AppointmentStatus;
+
+class AppointmentFilters extends QueryFilters
+{   
+    public function time($time)
+    {
+        $from = $time;
+        $to = $time + (request()->duration * 60000);
+        $checkBefore = 120 * 60000;
```
Comment (MohannadElemary2) [inline app/Filters/AppointmentFilters.php:13]: It's not a good idea to check if doctor is not having any appointments before the incoming appointment as he can have two or more following appointment without any problem.

Just check dates overlapping between the incoming appointment and the doctor current appointments:
https://stackoverflow.com/questions/325933/determine-whether-two-date-ranges-overlap

File: app/Filters/WorkingDayFilters.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{    
+    public function time($day)
+    {
+        
+        $day = $day/1000; 
+        return $this->builder->where('day', Carbon::createFromTimestamp($day)->dayOfWeek)
+            ->where(function ($query) use ($day) {
+                $query->where([
+                    ['from', '<=', Carbon::createFromTimestamp($day)->Format('h:m A')],
+                    ['from', 'like', '%' . Carbon::createFromTimestamp($day)->Format('A') . '%'],
```
Comment (MohannadElemary2) [inline app/Filters/WorkingDayFilters.php:16]: What is the need for this lines?
```
['from', 'like', '%' . Carbon::createFromTimestamp($day)->Format('A') . '%']
['to', 'like', '%' . Carbon::createFromTimestamp($day)->Format('A') . '%']
```

File: app/Http/Controllers/AppointmentController.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    protected $appointmentService;
+
+    public function __construct(AppointmentService $appointmentService)
+    {
+        $this->appointmentService = $appointmentService;
+    }
+
+    public function store(PatientReserveAppointmentRequest $request)
+    {
+        $appointment = $this->appointmentService->storeAppointment($request->except('status','cancel_reason'), $request->doctor);
```
Comment (MohannadElemary2) [inline app/Http/Controllers/AppointmentController.php]: Name it to `store`

File: app/Http/Controllers/AppointmentController.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+use App\Transformers\CreatedResource;
+use Illuminate\Http\Response;
+
+class AppointmentController extends Controller
+{
+    protected $appointmentService;
+
+    public function __construct(AppointmentService $appointmentService)
+    {
+        $this->appointmentService = $appointmentService;
```
Comment (MohannadElemary2) [inline app/Http/Controllers/AppointmentController.php]: It's a better idea to name the service as just `service`

File: app/Repositories/AppointmentRepository.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        // if ($appointment->count() == 1) {
+        //     abort(Response::HTTP_BAD_REQUEST, 'doctor has an appointment at this time');
+        // }
+   }
+
+   public function storeAppointment($data, $doctor)
+   {
+       $data['patient_id'] = auth()->user()->id;
+       $appointment = $doctor->appointments()->create($data);
+       return $appointment;
```
Comment (MohannadElemary2) [inline app/Repositories/AppointmentRepository.php]: This is a duplicated code with the `DoctorRepository`. Please just keep one of them and remove any duplicated code in the Pull Request

File: app/Services/AppointmentService.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        if ($approvedAppointments->count() > 0) {
+            $appointments = $approvedAppointments->map(
+                function ($item, $key) use ($data) {
+                    return (($item->time + ($item->duration * 60000)) > $data['time']);
+                }
+            );
+            ($appointments[0]) ? abort(Response::HTTP_BAD_REQUEST, "there is an already approved appointment at this time") : "";
+        }
+    }
```
Comment (MohannadElemary2) [inline app/Services/AppointmentService.php]: You already check that in the filter

Comment (mohamedebrahim2020) [reply app/Services/AppointmentService.php]: done

File: app/Repositories/AppointmentRepository.php (hunk 5 above)
Comment (mohamedebrahim2020) [reply app/Repositories/AppointmentRepository.php]: done

File: app/Http/Controllers/AppointmentController.php (hunk 4 above)
Comment (mohamedebrahim2020) [reply app/Http/Controllers/AppointmentController.php]: done

File: app/Http/Controllers/AppointmentController.php (hunk 3 above)
Comment (mohamedebrahim2020) [reply app/Http/Controllers/AppointmentController.php]: done

File: app/Filters/WorkingDayFilters.php (hunk 2 above)
Comment (mohamedebrahim2020) [reply app/Filters/WorkingDayFilters.php:16]: it`s ok

File: app/Filters/AppointmentFilters.php (hunk 1 above)
Comment (mohamedebrahim2020) [reply app/Filters/AppointmentFilters.php:13]: done

**Pull request merged**